workers/proto/kitchen_sink/kitchen_sink.proto

```diff
@@ -0,0 +1,328 @@
+  WorkflowInput workflow_input = 1;
+  ClientSequence client_sequence = 2;
+  // Technically worker options should be known as well. We don't have any common format for that
+  // and creating one feels overkill to start with. Requiring the harness to print the config at
+  // startup seems good enough for now.
+}
+
+// All the client actions that will be taken over the course of this test
+message ClientSequence {
+  repeated ClientActionSet action_sets = 1;
+}
+
+// A set of client actions to execute. Client actions are always executed concurrently, since
+// sequential execution can be represented by a series of 1-size action sets.
+message ClientActionSet {
+  repeated ClientAction actions = 1;
+}
+
+message ClientAction {
+  oneof variant {
+    DoSignal do_signal = 1;
+    DoQuery do_query = 2;
+    DoUpdate do_update = 3;
+  }
+}
+
+message DoSignal {
+  oneof variant {
+    // An signal handler must exist named `do_actions_signal` which is responsible for executing
+    // the action set.
+    ActionSet do_actions = 1;
+    // Send an arbitrary signal
+    HandlerInvocation custom = 2;
+  }
+}
+
+message DoQuery {
+  oneof variant {
+    // A query must exist named `report_state` which returns the `WorkflowState` message. The input
+    // is pointless and only exists to allow testing of variably-sized query args.
+    temporal.api.common.v1.Payloads report_state = 1;
+    // Send an arbitrary query
+    HandlerInvocation custom = 2;
+  }
+}
+
+message DoUpdate {
+  oneof variant {
+    // An update must exist named `do_actions_update` which does the same thing the signal handler
+    // would do when given the provided action set. The handler should return the `WorkflowState`
+    // when done.
+    ActionSet do_actions = 1;
+    // This update request must be rejected by the workflow using a handler named `always_reject`
+    google.protobuf.Empty reject_me = 2;
+    // This update request must be accepted and then failed by the workflow using a handler named
+    // `always_fail`
+    temporal.api.failure.v1.Failure fail_update = 3;
+    // Send an arbitrary update request
+    HandlerInvocation custom = 4;
+  }
+}
+
+message HandlerInvocation {
+  string name = 1;
+  repeated temporal.api.common.v1.Payload args = 2;
+}
+
+// Each workflow must maintain an instance of this state
+message WorkflowState {
+  map<string, string> kvs = 1;
+}
+
+message WorkflowInput {
+  repeated ActionSet initial_actions = 1;
+}
+
+// A set of actions to execute concurrently or sequentially. It is necessary to be able to represent
+// sequential execution without multiple 1-size action sets, as that implies the receipt of a signal
+// between each of those sets, which may not be desired.
+//
+// All actions are handled before proceeding to the next action set, unless one of those actions
+// would cause the workflow to complete/fail/CAN.
+message ActionSet {
+  repeated Action actions = 1;
+  bool concurrent = 2;
+}
+
+message Action {
```

Comment:
In this system how would I model activities or signal handlers that depend on each other or the main workflow thread?

Reply:
What does "depend on each other" mean exactly? A signal can set state, but not sure what the dependent signal is expecting. Are you asking for a wait-for-state?

Reply:
This is up to the generator. The generator can decide to make things depend on other things and that doesn't necessarily need to be expressible in the DSL - the generator keeps it in memory and may make use of the knowledge while the test runs.

Reply:
for example, signal A does not want to run past a certain point until the main workflow has run past a certain point.

I think wait for state would solve my ask.

Reply:
Yeah, wait for state will allow the generator to do arbitrary synchronization